src/main/java/io/vertx/core/net/impl/pool/SimpleConnectionPool.java

```diff
@@ -0,0 +1,781 @@
+      if (slot.concurrency() > 0) {
+        return slot;
+      }
+    }
+    return null;
+  };
+
+  /**
+   * A slot for a connection.
+   */
+  static class Slot<C> implements PoolConnector.Listener, PoolConnection<C> {
+
+    private final SimpleConnectionPool<C> pool;
+    private final EventLoopContext context;
+    private final Promise<C> result;
+    private PoolWaiter<C> initiator;
+    private C connection;    // The actual connection, might be null
+    private int index;       // The index in the pool slots array
+    private int concurrency;    // The current concurrency, i.e the number of acquisitions this connection supports
+    private int maxConcurrency; // The connection maximum concurrency
+    private int capacity;      // The connection capacity
+
+    public Slot(SimpleConnectionPool<C> pool, EventLoopContext context, int index, int capacity) {
+      this.pool = pool;
+      this.context = context;
+      this.connection = null;
+      this.concurrency = 0;
+      this.index = index;
+      this.capacity = capacity;
+      this.result = context.promise();
+    }
+
+    @Override
+    public void onRemove() {
+      pool.remove(this);
+    }
+
+    @Override
+    public void onConcurrencyChange(long concurrency) {
+      pool.setConcurrency(this, concurrency);
+    }
+
+    @Override
+    public Context context() {
+      return context;
+    }
+
+    @Override
+    public C get() {
+      return connection;
+    }
+
+    @Override
+    public int concurrency() {
+      return concurrency;
+    }
+
+    @Override
+    public int maxConcurrency() {
+      return maxConcurrency;
+    }
+  }
+
+  private final PoolConnector<C> connector;
+  private final int maxWaiters;
+  private final int maxCapacity;
+  private final int[] capacityFactors;
+  private final Executor<SimpleConnectionPool<C>> sync;
+  private final ListImpl list = new ListImpl();
+
+  // Whether the pool is closed
+  private boolean closed;
+
+  // Selectors
+  private BiFunction<PoolWaiter<C>, List<PoolConnection<C>>, PoolConnection<C>> selector;
+  private BiFunction<PoolWaiter<C>, List<PoolConnection<C>>, PoolConnection<C>> fallbackSelector;
+
+  // Connection state
+  private final Slot<C>[] slots;    // The pool connections, this array is not sparse
+  private int size;                 // The number of non null slots
+  private int capacity;             // The pool capacity
+
+  // The queue of waiters
+  private final Waiters<C> waiters;
+
+  SimpleConnectionPool(PoolConnector<C> connector, int[] maxSizes) {
+    this(connector, maxSizes, -1);
+  }
+
+  SimpleConnectionPool(PoolConnector<C> connector, int[] maxSizes, int maxWaiters) {
+
+    int[] capacities = new int[maxSizes.length];
+    int maxCapacity = 1;
+    int numSlots = 0;
+    for (int i = 0;i < maxSizes.length;i++) {
+      int maxSize = maxSizes[i];
+      if (maxSize < 1) {
+        throw new IllegalArgumentException();
+      }
+      maxCapacity *= maxSize;
```

Comment:
Not sure if I understand here, if 
`Http1xConnection: {max-pool: 5, capacity: x}`, `Http2Connection: {max-pool: 2, capacity: y}`, should the `maxCapacity = 5x + 2y` ?

According to your equation, we get `maxCapacity = 5 * 2 = 10, http1x.capacity = 2, http2.capacity = 5`. added them up again, we get `maxCapacity = 20` if above sentence is correct. :)

Reply:
what we can get at most is:

- 2 HTTP/2 connections (5+5)
- 5 HTTP/1 connections (2+2+2+2+2)
- 1 HTTP/2 connection + 3 HTTP/1 connections (5+2+2+2)